servers/mcp-neo4j-cypher/src/mcp_neo4j_cypher/server.py

```diff
@@ -1,164 +1,155 @@
+    ),
+) -> list[types.TextContent]:
+    """Execute a read Cypher query on the neo4j database."""
+
+    if _is_write_query(query):
+        raise ValueError("Only MATCH queries are allowed for read-query")
+
+    try:
+
+        async def read(tx: AsyncTransaction, query: str, params: dict[str, Any]):
+            raw_results = await tx.run(query, params)
+            results = [r.data() async for r in raw_results]
+            return results
+
+        async with neo4j_driver.session(
+            database=os.getenv("NEO4J_DATABASE", "neo4j")
+        ) as session:
+            results = await session.execute_read(read, query, params)
+
+            logger.debug(f"Read query returned {len(results)} rows")
+
+            return [types.TextContent(type="text", text=str(results))]
+
+    except Exception as e:
+        logger.error(f"Database error executing query: {e}\n{query}\n{params}")
+        return [types.TextContent(type="text", text=f"Error: {e}\n{query}\n{params}")]
+
+
+@mcp.tool()
+async def write_neo4j_cypher(
+    query: str = Field(..., description="The Cypher query to execute."),
+    params: Optional[dict[str, Any]] = Field(
+        None, description="The parameters to pass to the Cypher query."
+    ),
+) -> list[types.TextContent]:
+    """Execute a write Cypher query on the neo4j database."""
+
+    if not _is_write_query(query):
+        raise ValueError("Only write queries are allowed for write-query")
+
+    try:
+
+        async def write(tx: AsyncTransaction, query: str, params: dict[str, Any]):
+            return await tx.run(query, params)
+
+        async with neo4j_driver.session(
+            database=os.getenv("NEO4J_DATABASE", "neo4j")
+        ) as session:
+            raw_results = await session.execute_write(write, query, params)
+            counters = raw_results._summary.counters
+
+        logger.debug(f"Write query affected {counters}")
+
+        return [types.TextContent(type="text", text=str([counters]))]
+
+    except Exception as e:
+        logger.error(f"Database error executing query: {e}\n{query}\n{params}")
+        return [types.TextContent(type="text", text=f"Error: {e}\n{query}\n{params}")]
+
+
+def healthcheck(neo4j_driver: AsyncDriver) -> None:
+    """Confirm that Neo4j is running before continuing."""
+
+    attempts = 0
+    success = False
+    print("\nWaiting for Neo4j to Start...\n")
+    time.sleep(3)
+    while not success or attempts <= 3:
+        try:
+            with neo4j_driver.session() as session:
+                session.run("show databases")
+            success = True
+        except Exception:
+            attempts += 1
+            print(
+                f"failed connection {attempts} | waiting {(1 + attempts) * 2} seconds..."
+            )
+            time.sleep((1 + attempts) * 2)
+    if not success:
+        raise DatabaseError()
+
+
+def main() -> None:
+
+    # Test the connection
+    healthcheck(neo4j_driver)
+
+    mcp.run(transport="stdio")
```

Comment:
we should also allow SSE and other transports (Streaming) based on entrypoint or something?